### ISSUE TITLE
What is our git development model?

### ISSUE THREAD

**Issue by laserson** [airr-community/airr-standards#35]
If we're going to do most development on the `development` branch, could we make it the default branch when you land on the github page?

(In case it's still up for discussion, I'd vote to do development on master, though)

**Comment by schristley** [comment]
My opinion is that people reading the MiAIRR manuscript, and going to the GitHub page, should by default see the latest released version of MiAIRR. And if they clone the repository, the master (or default branch), also by default should be the latest released version. All development, intermediate changes, and etc., should be kept on a separate branch so as not to confuse people.

That's just my preference though. This hasn't been "decided" in any way, and we certainly should discuss in a larger context. My underlying concern is that we don't have a coordinated documentation design for the AIRR Community. You will note that no URL was provided in the MiAIRR manuscript, just an email address. The [AIRR Community website](http://www.airr-community.org) is just a redirect to SFU, and we don't have much control over that. Do we use readthedocs even though it is somewhat python oriented, and may not be the best way to describe all docs for the AIRR Community? Are we using DOI's created on [zenodo](https://doi.org/10.5281/zenodo.997605) for releases? Should we setup our own web server? Right now it feels that GitHub is the current source of truth, and people are going to get directed here for docs, templates, and etc., until we have an overall strategy in place.

**Comment by laserson** [comment]
I find it very surprising to land on a GitHub page for a repository and not see the latest commit for the latest development.  It's quite unusual for a GitHub page to point to a release.  Consumers/users of the AIRR standards should view the published documentation (soon to be at docs.airr-community.org) and install released software using common package managers (e.g. pip/PyPI, CRAN, Maven, etc).  People should really only be going to the GitHub page if they're interested in contributing to development, to report bugs, or to see what's been going on recently.

**Comment by schristley** [comment]
That's understandable. I happen to be the other way in that I'm surprised to land on a repository page and not be immediately presented with a stable release. But each organization has a different dev workflow. My background is often in tools that don't use the package managers that you mention, often instead being in the distribution package (apt-get, rpm, yum), which means they are often out-of-date, but it is interesting that some popular languages have done an end-around this bottleneck and created their own package systems. It is good that we will have a dedicated doc website.

**Comment by schristley** [comment]
For the workflow you are suggesting, how are multiple versions handled and maintained? I'm seeing the definite scenario that there will be some fine-tuning of the 1.0 spec for the production systems out there, which will require some point releases, but we don't want that intermixed with the 2.0 work.

**Comment by laserson** [comment]
This is basically what I'm most familiar with:

https://barro.github.io/2016/02/a-succesful-git-branching-model-considered-harmful/

To summarize:
- One branch where development evolves, e.g., "master" or "trunk" or "devel" etc.  Everything gets committed into this branch
- Individuals/groups can work on feature branches. When they want to submit their work, they rebase it on the latest master and submit a pull request.  They can also rewrite their own history a bit in the process to clean up/squash their commits.
- The master branch should never be broken.
- Releases are basically just a tag on a particular commit. Typically, you'd also have a branch for that release as well.  You'd follow your release process at that commit to push artifacts to any repositories like PyPI, CRAN, etc.
- If you need to patch a particular release (e.g. for a bugfix), you'd add a commit off of the release branch.  Oftentimes, you find a bug that you fix on the master branch, but you also want to apply it to a previous release.  In that case, you'd commit to master and then cherry pick that commit onto the release branch as well.

There's a good diagram in the link above.

**Comment by schristley** [comment]
That's fine, I don't have any particular issue with it. I consider the `development` branch an additional abstraction that's helpful in some situations, but it's not necessary.

I agree that the master branch should never be broken, but technically this won't happen with this model, unless we strictly enforce a policy. For example, Christian updates the MiAIRR, but he's not going to update the python code, nor the R code, nor maybe some for the docs, etc. All of that will need to be done before the pull request can be merged. I'm okay with that, we just need to enforce the policy. Even then it's unclear whether we could attain 100% coverage, for example testing submission templates to repositories, testing the API for web services, etc.

In my model the master branch is never broken (except for unknown bugs), but the development branch may for periods of time (think integration of features/components). With our small group and tight coordination, these situations may not arise.

The important part to me is handling the release branches in parallel, so long as we agree that when v1.0 comes out, it goes on its own branch, and `master` then becomes v2.0 work. In my model, there are release branches for major versions, so no change there.

Individual/groups working on features for v1.0 would need to know to create their feature branch from the v1.0 branch, and not master. In fact, we can get rid of `master` as the name and use the version number for the branch names to be more explicit.

**Comment by BusseChristian** [comment]
I do not have to much experience with these models in "real" (i.e. distributed and asynchronous settings), so no strong opinions, beside that I prefer `rebase`d histories. So with `development` gone and `master` locked, would the usual commit cycle look like below?
- fetch `master` from remote
- new local branch from `master`/`HEAD`
- edits & testing
- fetch again `master` from remote
- [if necessary, rebase to current `master`/`HEAD` locally]
- push new branch to remote
- pull request

**Comment by schristley** [comment]
Yep, pretty much, the key point is to put your work on its own branch and submit a pull request when ready to merge.

Is there a way to get the consistency check travis job to run on a branch without doing a pull request? I have settings enabled that those checks must be passing in order to do a merge, but if it's a hassle to run them locally, would be nice to just run them on a branch.

**Comment by laserson** [comment]
I think that Travis currently build every pull request and every branch that's pushed into the repo.

**Comment by laserson** [comment]
Also, perhaps we can disable merge commits?

![screen shot 2017-11-06 at 13 11 59 pm](https://user-images.githubusercontent.com/118141/32456637-522d112e-c2f4-11e7-8aa3-971ae14e717f.png)

You can see how the PRs I submitted should've been clean linear commits added to the history, but the merges made it look complex.

**Comment by schristley** [comment]
I admit I've never paid much attention to these different styles of merging, as they weren't available before git. When I merged your two pull requests, the second wanted me to update with the master branch as it was changed, which I guess explains what you are seeing?

I notice (from airr-formats work) that all of the commits got "collapsed". I tend to do individual commits with messages for specific files. I personally like to see that history but that all gets collapsed into a combined commit message. Is there any way to keep that message/file relationship yet have the linear history that you want?

**Comment by schristley** [comment]
Ok, cool, then in that case go ahead and change the settings appropriately.

**Comment by BusseChristian** [comment]
How and where do we best summarize these decisions (i.e. the current development model)?
A `CONTRIBUTE.md` in the main directory?

**Comment by schristley** [comment]
For [VDJServer](https://bitbucket.org/vdjserver/vdjserver-web-backbone) we have section in the `README.md` that describes the development guidelines, so that it's immediately available from the main page.

**Comment by schristley** [comment]
I noticed there is a [contributing](https://github.com/airr-community/airr-standards/community) section in the settings, maybe that is the place to put it.

**Comment by schristley** [comment]
Now that we are one big happy repo, how about we create a set of labels (or projects?) so we can tag the different groups (MiAIRR, formats, common repository, etc.)?

**Comment by javh** [comment]
I don't know how I missed this issue, but I'm full of opinions! I'm pretty much in full agreement with @schristley.  I would prefer something like a simplified [Gitflow approach](https://www.atlassian.com/git/tutorials/comparing-workflows/gitflow-workflow).

In our repos we use a simple workflow where `default` (`master`) is used development, you don't commit broken code to `default`, and if you need to break something you use a feature branch. The reasons we do it this way are:

+  We're always working with the latest code, so `develop` and `default` might as well be the same. We would never use `default` internally if it was only for releases.
+ Repos don't have subprojects. Every repo is a standalone package so there's only one development path.
+ And the big reason: it's easier to onboard new people who join the lab. The workflow needs to be accessible to 1st years who are unfamiliar with software development.

But, I don't think any of these criteria apply to the airr-standards repo. I'm thinking something like the following would be better:

+ `master` should be stable code only with development occurring in the `develop` branch.
+ Merges into `master` are done via pull requests from `develop`.
+ Small changes can be committed directly to `develop`, but larger changes should be a feature branch off `develop`. Feature branches should never spawn from `master`, but new development branches might.
+ Anonymous changes would use pull requests, but work by the core developers can just be commits to `develop`.

I'm thinking this would be a lot easier to work with. 

There wouldn't be all this pull request approval/merge overhead, which currently means someone has to dedicate a chunk of their time to micromanaging the repo. Whomever deals with merges only has to deal with it when you want to pull stable code from `develop`.

There's no lag for those actively working on the code. For example, we recently had the sphinx doc pull request sitting there for 12 days. Which means if I wanted to update the docs, I couldn't really do so without stacking another pull request on top of a pull request. I know this was because of the AIRR meeting, but we're not always going to have compatible schedules.

This would simplify doc maintenance too, because then we can just have the ReadTheDocs site land on `stable` without having to update the landing/building version in ReadTheDocs with each release. We can just map `stable` -> `master` and `latest` -> `develop`.

If we wanted, we could have project specific `develop` branches. We could have `develop-formats-py`, `develop-formats-r`, `develop-miairr`, etc. Provided these are associated with a strict directory structure, then we can have development proceed independently within separate `develop-*` branches. Merging into master only when a project hits a stable release with meta-versioning on the `master` branch. New `develop-*` branches could be spawned for new projects as needed.

We're a small group. Seven contributors so far? With 3-4 people actively contributing code per project? I'm sure we'll get more contributors over time, but I don't think we need a formal pull request and code review process for such a small group. Just being able to commit would be a lot simpler/faster.

One downside about github is that it doesn't notify you of commits to watched repos. So without the pull requests you fail to notice commits. Though, we could setup a slack channel with github integration if we wanted to work around that.

That's my $0.02. Or maybe it was $0.08.

**Comment by laserson** [comment]
I fully support your generally preferred approach, @javh, and I think it would be appropriate here too.  What you actually propose for this repo I think will actually just add overhead.

Basically, there needs to be one branch where new stuff goes.  And bc we'll have versioned releases of the standards, there need to be branches for them.  And that's basically it.  The branch where the new stuff goes can be called "develop" or "master" or whatever, but there is really no need for a "develop" and a "master" if we'll also have branches for the versions, which we will.  (I'd personally prefer "master" since it's more common.) People can choose to privately work on their own branches in prep for a pull request, or they can work directly on master for their own fork, but it doesn't really make a difference.  Having multiple develop branches also seems overcomplicated.

Also, the pull request/merge overhead is independent of the git model we choose.  It depends on whether we want literally every commit to be +1'd by one of the "committers", which is typical but also a lot of overhead at this point. A real question is whether we want to let committers commit code without getting approval from someone else. To your point, with just a few active committers at the moment, we can probably get by with the policy "don't do something stupid or piss people off" for now, and also give the core devs commit rights.

Regarding the git model, I'd just reiterate that [gitflow complicates things unnecessarily IMO](https://barro.github.io/2016/02/a-succesful-git-branching-model-considered-harmful/).

> We can just map `stable` -> `master` and `latest` -> `develop`.

This seems like it would be problematic potentially. When people actually use the airr-standard, we want to make sure they're using a particular version.  If they just point to "stable" then a new version release on our end could break their code/repo.

> One downside about github is that it doesn't notify you of commits to watched repos. So without the pull requests you fail to notice commits.

We could enforce the policy that all commits go through a pull request, but let you accept the PR yourself, at least for the core group of committers. This would add a small amount of overhead, but would at least generate the notifications.

---

So overall, I'm advocating:

- a single branch for development ("master") and release branches that correspond to released versions; linear history (no "merge" commits); "master" should never be broken (once we get there)
- all contributions to the master branch through PRs, either through forked repos or feature branches (the latter only working for people with write access to the repo, the former for any contributor)
- core group of devs that can commit code (accept PRs), including their own
- for now, have a policy of "don't do stupid shit"; if you think your change might be controversial, wait for approval from other committers; we can revisit this if the dev group starts to get bigger

**Comment by schristley** [comment]
We don't have an actual release branch, in my mind that's confusing matters. Maybe we are all talking the same thing but are mixing the words. In the gitflow, "master" is the release branch and "develop" is where development occurs, "master" is never broken but "develop" may be as integration work is performed. In Uri's model, "release" is the release branch and "master" is where development occurs. The "release" branch should never be broken, but "master" may be as integration work is performed.

So are we all talking the same thing, or am I still misunderstanding something?

**Comment by javh** [comment]
Hrm. Okay, so this seems to depend upon whether you want to handle releases as branches or tags. In the simplest case we are discussing two options:

1. Development in `master` with releases branches off `master`. Release branches are just a single changeset.
2. Development in `develop` with releases as tags in `master`. Master is just one changeset per release.

We do the first in our repos, except releases are just tags instead of branches. Which is another option. The second gives us more flexibility to do long lived development branches if we want, but we don't have to.  With just a single `develop` branch, they are really similar. I think we could do either and it would feel pretty much the same day-to-day. However, though the first would remove some flexibility and the ability to manage releases through pull requests. Also, it would mean having unstable code in release branches. For example, if the python library hits a milestone, but the R library is still a work-in-progress, then the unfinished R library will be tagged with the python library version.

**Re: ReadTheDocs**

> This seems like it would be problematic potentially. When people actually use the airr-standard, we 
> want to make sure they're using a particular version. If they just point to "stable" then a new version 
> release on our end could break their code/repo.

I don't think so. We'll want the docs to land on whatever the latest release version is. If the tip of 'master` is always the current release, then that's the same as manually setting the landing page to the docs that are build from the most recent release branch. I think this again comes done to whether `master` is used only for releases or not. If `master` is stable but not released code, then, yeah, that would be confusing.

**Re: Everything through pull requests**

I think this is likely to be my biggest issue. I think requiring approval from someone else for each commit is going to be burdensome. But if you can just approve your own pull requests, then why bother with pull requests instead of just doing a normal commit?

**Comment by schristley** [comment]
And on a separate note, can somebody explain this "no merge commit" thing. I've seen it mentioned multiple times. For pull requests, sometimes they are old and it says that it is "out of date with the master". There is an `update branch` button (it's called sync on bitbucket). I click that button all the time. Should I be doing something else?

**Comment by schristley** [comment]
> handle releases as branches or tags

Yeah I might have confused matters there because I tend to think of release as tags sitting on the "release" or "master" branch, i.e. what Jason indicates.

> 2. Development in develop with releases as tags in master. Master is just one changeset per release.

and so the current repo is a mixture. Instead of making a 0.1 tag on master, I probably should have made a release-0.X branch and put the tag on it.

**Comment by javh** [comment]
Oh, and for the simple case we'll still have the "unstable code in release branches/tags" no matter which way we go. I think we'd need the more complex multiple develop branches to fix that. If we care.

**Comment by schristley** [comment]
> Also, it would mean having unstable code in release branches. For example, if the python library hits a milestone, but the R library is still a work-in-progress, then the unfinished R library will be tagged with the python library version.

Ugh. I don't like that at all, but it seems we have to deal with it. My hope was that everything would be properly working before we did a release. This will only become more complicated as we start creating specs/APIs for the common repository and germline WGs.

**Comment by javh** [comment]
> Ugh. I don't like that at all, but it seems we have to deal with it. My hope was that everything would 
> be properly working before we did a release. This will only become more complicated as we start
> creating specs/APIs for the common repository and germline WGs.

Yeah, we decided to put multiple projects in the same repo. I'm not seeing a way around the problem except some sort of system that somehow divides the repo into sub-repos for each project (eg, multiple develop branches).

I mean, we could just wait on releases until every project in the repo hits a stable milestone. That's all I can think of. It might be annoying though. Maybe someone else has a better idea?

**Comment by javh** [comment]
Sorry for the spam, but I just thought of another upside to @laserson's proposal. If we do each release as a separate branch, then it'll make maintenance releases easier. For example, if we discover a bug long after a release and the affected code is no longer in master, it's easy to just continue the original release branch to implement the fix.

**Comment by laserson** [comment]
I think we are getting confused about terminology.  Ultimately, there is simply a single tree of commits.  A branch and a tag are essentially the same thing...it's just an object that points at a commit on the tree.  The advantage of actually using a branch vs a tag in git is that you can continue adding commits to the branch, in case we need to do a bugfix on an older release (e.g., bump the "z" in version "x.y.z"). But in reality, we'd just do both.  Tag a commit on release, and add a corresponding branch as well.

Therefore, for the docs, we can have a "stable" branch which is the landing page, but it should not point at the "master" branch, but rather a branch that's labeled with the latest version.  If the version branch and the master branch both point at the same commit (i.e., at the tip of the tree), that's no problem. But any time someone uses our software or standards, they should always be referencing a particular version. They can reference "master" if they want, but then they have to deal with the floor falling out beneath them. Part of our release process would be to update the "stable" pointer to the latest released version.

More significantly, having a develop and a master branch just means there are two tips of the tree that are moving forward, and they actually diverge quite a bit before they're merged together again periodically. This seems redundant and potentially confusing.  I think there should be just one tip for the mainline of development, and it should never be broken.  You can feel free to create a long-lived feature branch and have multiple people commit to it for a while before incorporating it into "master".

> [On multiple repos vs one repo:] My hope was that everything would be properly working before we did a release.

That's why I'd advocate that we actually stay on v0.x rather than rushing to v1.0. Too late to go back to this?

One option is to keep the `airr-standards` repo just for "standards" and have separate python and R repos.  Both approaches have advantages and disadvantages.  I slightly lean towards one repo for the moment.

> But if you can just approve your own pull requests, then why bother with pull requests instead of just doing a normal commit?

Purely for record-keeping purposes. It also lets people comment on your changes. (Even for stuff already committed I believe.) If you can accept your own pull requests, it should be pretty small overhead. And if we eventually change the policy where we want to have someone else approve, it'll be super easy.

> if the python library hits a milestone, but the R library is still a work-in-progress, then the unfinished R library will be tagged with the python library version.

There's just one version for airr-standards. "master" should never be broken, so whenever you tag a release, everything should continue to work. If there's not much change in the R lib but a lot of change in the python lib at a release, doesn't really make a difference.

> I think this again comes down to whether master is used only for releases or not.

Of necessity, "master" or "develop" or whatever is always "used" for a release, bc whatever commit you tag as the release was at one point tip of the tree. Which is why there is no point in having both a "develop" and a "master".

> another upside to @laserson's proposal. If we do each release as a separate branch, then it'll make maintenance releases easier. For example, if we discover a bug long after a release and the affected code is no longer in master, it's easy to just continue the original release branch to implement the fix.

Yes! This is exactly the reason for doing a branch instead of a tag.

**Comment by javh** [comment]
> More significantly, having a develop and a master branch just means there are two tips of the tree 
> that are moving forward, and they actually diverge quite a bit before they're merged together again 
> periodically. This seems redundant and potentially confusing. I think there should be just one tip for 
> the mainline of development, and it should never be broken. You can feel free to create a long-lived > feature branch and have multiple people commit to it for a while before incorporating it into "master".

So I think this is a point of confusion. Master and develop aren't independently moving forward. Develop is always equal to or ahead of master in the simple linear case. The difference is just whether branches "go out" from develop (master) to release (trunk to leaves) or "go in" from develop to release (master) (roots to trunk). By definition, if we went with master as a single release branch, the tip of master would be the latest stable release.

After talking about it, it sounds like we're in favor of a simple linear workflow, yes? So in that case, it doesn't really matter to me whether we use branches for releases or master for releases. If go with the former, we lose the ability to do complex non-linear things like having independent versions for each sub-project in the repo. If we're all okay with that downside, then I don't think it makes much difference. 

I'm cool with the cactus model then, because `develop` in Gitflow and `master` in the cactus model should behave the same. 

I'm still not certain it's wise to have the landing page for the github repo being unstable code, but if we properly maintain the README/docs then it should probably be okay. If we don't regularly update the version in the README/docs people will come to the page, read version 1.0, but end up downloading the unstable tip.

> Purely for record-keeping purposes. It also lets people comment on your changes. 
> (Even for stuff already committed I believe.) If you can accept your own pull requests, it should be 
> pretty small overhead. And if we eventually change the policy where we want to have someone 
> else approve, it'll be super easy.

You can comment on commits directly. You can even comment on individual lines of code in a commit. I think commits are sufficient record keeping if you don't need to discuss the changes before merging.

**Comment by schristley** [comment]
Thanks everybody for their effort in discussion, I think we are converging.

> That's why I'd advocate that we actually stay on v0.x rather than rushing to v1.0. Too late to go back to this?

I think so, yes. At the AIRR meeting, MiAIRR was presented as v1.0, with goals for v1.1 and longer term goals for v2.0. We could try to make some technical argument that the repo's version are different from MiAIRR but that will likely just confuse people versus being helpful. 

We got agreement that submissions (e.g. to NCBI) will have MiAIRR version numbers associated with them, though the technical implementation needs to be resolved. This means that we need to consider release branches as living in perpetuity just like the data. While it might be nice to think that V2.X could interoperate with V1.X data, there's no guarantee of that.

**Comment by schristley** [comment]
> But if you can just approve your own pull requests, then why bother with pull requests instead of just doing a normal commit.

I like the pull requests because it runs the travis job to insure some gross inconsistency wasn't introduced. If you are on a branch, you can see the travis results though they aren't obvious, you have to go to travis and click around. If you commit directly to master, you aren't going to get the check beforehand.

Now you can argue that developers could run the consistency check themselves. But I've tried to run the consistency checks in my development environment, talk about a total pain in the butt, I had to setup a whole docker image, blah blah... So it's nice to let GitHub do it for me.

However, I'm in agreement to keep things simple. I was initially trying to setup all of these formal rules for `master` to insure it doesn't get messed up in relation to MiAIRR releases, but I think we've resolved that, so we can simplify down to 2 steps:

1. Always create a pull request.
2. Pull request must pass the travis job checks to merge.

Now `owners` can force merges even though the checks fail, and they can also commit directly to master without a pull request, so we just have to agree `don't do that!`

**Comment by schristley** [comment]
Okay, to get some closure on this. I've created a document for the development model. Please review pull request 72.

**Comment by javh** [comment]
**Pull Requests**

> I like the pull requests because it runs the travis job to insure some gross inconsistency 
> wasn't introduced. If you are on a branch, you can see the travis results though they aren't 
> obvious, you have to go to travis and click around. If you commit directly to master, you 
> aren't going to get the check beforehand.

That is nice, but I personally don't think it's worth the hassle of having to deal with multiple local repos and merges. It also strikes me as somewhat non-linear. Every time you create a fork it's like an anonymous branch, and if the repo changes between your fork and your pull request, then you have to deal with resolving it. Whereas, if you're just working with a local sync of the repo you can make your changes, fetch, commit and push, thus avoiding merges.

I'm pretty strongly opposed to having to do everything through pull requests. But, here's a thought... we could just not set a rule about it. If you want to do a PR for the remote consistency check you can, or if you want to do pushes and do local consistency checks you can do that too. Just make a judgment call on a case-by-case basis. In the end, it all just ends up as commits in the history regardless of whether it originates as a push or pull request. 

This is what we did in the airr-formats repo without ever discussing it and I don't recall any issues. Maybe some confusion on whether to rebase branches or not? I guess we should talk about that too...

**Versioning**

I think we need to think about this carefully. Off the top of my head, I can only think of two mostly bad options (below). Anyone have any other ideas?

**1)** We meta-version the repo and wait for each subproject to hit a stable version before incrementing the meta-version.  For example, we might have the following meta-version with component versions described:

AIRR Standards v1.0:
+ MiAIRR v1.0
+ PyAIRR v0.1

The downside being that this is more overhead. It's not huge overhead, but the more little tasks you add to each release, the easier it is to forget one. Also, the meta-version would really be more of a convenience for us than something users would refer to, which is potentially confusing.

**2)** We version the repo with named subproject versions that denote which subproject hit release for the given change set.  For example, we could label the repo `MiAIRR v1.0` now. When the python library is done, then we label the repo `PyAIRR v0.1`. There's no clear linear version history, but you would know which changeset denotes a release for the given project.

**Workflow**

I feel like if we want input from others a visual aid might be helpful at this point. Essentially, we are debating between the workflows shown below. Each colored circle is a release and the arrow is the development branch.

![workflows](https://user-images.githubusercontent.com/4105629/33967938-984859c8-e033-11e7-8551-93dfbdd5546a.png)

**C** is what we do for our repos. **A** is what @laserson is suggestion. **B** is what @schristley and I proposed.

At this point, if I understand our goals correctly, @laserson has convinced me **A** will work if we're smart about the versioning. It's closer to what I'm used to anyway, so I'm not going to complain if that's the decision. :)

**Comment by laserson** [comment]
(terse bc on my phone at airport)

I still don't understand why we can't have a single version for the whole repo. If a new version only updated the python code, the new version number will still make it clear what's on the R code. And given version bump in any system will almost always focus on some subset of features.

Regarding your diagrams, Jason, A and C start off the same. You only get the extra branches coming off in A if there are actually updates that need to be made to old releases. (Actually, unless you do A, I'm not sure how to even update an older release.)

**Comment by laserson** [comment]
Also, I'm not sure how B works. If I understand correctly, each colored node is a release, so how does information get transferred from the main trunk into a release?

**Comment by laserson** [comment]
One of my main points is that different lines of development always need to be merged eventually. Either you do it with a merge commit, which creates the parallel histories, or you do that by rebasing your branch on the tip of master, which results in a linear history.

Finally, @javh, it is not necessary to fork a repo to submit a pull request. You can create a PR from a branch in the same repo.

**Comment by javh** [comment]
In order:

1. Let's say you have two packages in a repo. The code for Package A is ready for release at version 1.1. The code for Package B is currently incomplete and is somewhere between version 2.1 and 2.2. What do you tag the repo as? And if someone downloads the versioned changeset, thinking the code base is stable because there was a release annotation, then they get a broken version of Package B.
2. **A** is branched releases and **C** is tagged releases. Same general idea, but **C** is just annotating changesets.
3. This is the workflow where you merge from `develop` into `master` when you want to do a release, so the releases are tags on `master` (`master` has the circles on it).
4. Well sure, you can make a bunch of branches, but then aren't we doing Gitflow at that point? If we can push to branches without pull requests, then we pull request into master, then we don't have a single development branch anymore, right? I'm not saying we shouldn't do feature branches when appropriate, I'd just really rather not do a fork/branch for every little thing.

**Comment by schristley** [comment]
> 1. Let's say you have two packages in a repo...

The sub projects cannot have independent versioning, otherwise users won't know what goes with what. What version of MiAIRR is supported by this tool version? Given some data in a specific MiAIRR version, what tool version do I need? With everything as the same version then there is no ambiguity.

I realize that normally this could cause all kinds of problems, but I think we are okay because MiAIRR is stable and will change very infrequently. Most of the stuff we are debating are internal design decisions that won't affect what the average AIRR user sees. Plus I think we can control this by which versions we stick into the public package managers (pip, cran, etc.). Say for example, v1.0.1 python is ready but R isn't, well we put a python package on pip, but don't put anything on cran. Later R is ready, in v1.0.2, so we put it up on cran. Now technically, there is a v1.0.2 of the python package, we can decide if we want to put it or not depending if anything has changed.

If we make a major internal re-design decision, like changing the format of the specification files, then we just have to be conservative and make sure everything is updated before we do a release, and not rush out one tool just because it is ready.

I think if there is some vitally important project, that needs those changes in some stable way, then we can do something ad-hoc like a special tag on the master branch, e.g. python-dev-v1.0.5

**Comment by schristley** [comment]
I'm not sure that **B** was what I'm thinking, probably because I don't see changes on the development branch flowing into the release branch, but anyways no matter, I'm good with **A**. **C** is not flexible enough for us, so I'd vote for **A**, however with a minor clarification. I don't think every release should be it's own branch, I think every major release should be it's own branch. So there is a `release-1.x.x` branch, a `release-2.x.x` and so on.

**Comment by schristley** [comment]
> I'm pretty strongly opposed to having to do everything through pull requests...

I understand that, and I'm not strongly set on requiring pull requests. What I don't like is allowing anybody to commit without a PR, but I'm cool with the idea that some people are allowed. With the knowledge that if you screw it up, you owe us all a round of drinks at the next AIRR meeting ;-D

And some things really do need to be off limits, i.e. don't go changing the TSV file or the specifications files, until other people have had a chance to review. Simple bug fix to some code, no problem.

**Comment by javh** [comment]
### Workflow

Well, I feel like we've been talking about something like **D** occasionally (I updated my visual aid above), which would be having merges into master from development branches/forks, with master containing unstable code, and releases as branches from master. Which I think is more complex than necessary - basically combined **A** and **B** into something with messy parts of both.

To answer @laserson's question from earlier about how you update old releases in **C** (and **B**)... you don't. Versions should be linear. If you really need to, then you'd have to spawn a maintenance branch from the version tagged changeset, but that would be a special case and not standard practice.

And, yeah, we can set some guidelines for what kind of changes should get reviewed. I'm certainly not envisioning global write access either. I've always felt that the entire point of pull requests was access control - providing a mechanism for people without write access to make commits.

**Comment by schristley** [comment]
> we've been talking about something like **D** occasionally

Yes exactly, though **D** is making feature/development branches more explicit, I kinda assumed those were there regardless. I pretty much always put major development on a branch.

What isn't clear to me is what those colored release dots on **D** are exactly, are they v1.X, v2.X, v3.X? Or are they v1.1.X, v1.2.X, v1.3.X, or are they v1.1.0, v1.1.1, v1.1.2? For releases and version, I think more of that branch in **B** with the color dots. In which the branch represents v1.1.X, and each of the dots would be v1.1.0, v1.1.1, v1.1.2. When the spec updates to v1.2, we create a new release branch for v1.2.X releases.

**Comment by javh** [comment]
> Yes exactly, though D is making feature/development branches more explicit, I kinda 
> assumed those were there regardless. I pretty much always put major development on a branch.

They are. Any of them could have feature branches, but I didn't want to make the diagrams overly complex. It's just an outline of the day-to-day practice.  In **D**, this would be always doing your development in a branch/fork and then having to do a merge into master (ie, requiring pull request for everything).

> What isn't clear to me is what those colored release dots on D are exactly, are they v1.X, v2.X, 
> v3.X? Or are they v1.1.X, v1.2.X, v1.3.X, or are they v1.1.0, v1.1.1, v1.1.2? For releases and version, I > think more of that branch in B with the color dots. In which the branch represents v1.1.X, and each > of the dots would be v1.1.0, v1.1.1, v1.1.2. When the spec updates to v1.2, we create a new release > branch for v1.2.X releases.

The dots are just "versions" - whatever that means. If we want to do branches for major versions and tags for minor versions, I think that's maybe something to sort out later. Personally, I think using a different mechanism for different types of versions is overly complicated. But I can see some logic to it.

**Comment by schristley** [comment]
> The dots are just "versions" - whatever that means. If we want to do branches for major versions and tags for minor versions, I think that's maybe something to sort out later. Personally, I think using a different mechanism for different types of versions is overly complicated. But I can see some logic to it.

Okay, though later is pretty soon. We already should have a v1.0 MiAIRR, and I'd actually like to release the python package as soon as possible. iReceptor and VDJServer are both pushing for a new public release soon (maybe end of Jan), and they should really point to released versions.

Anyways, the issue I feel we need to consider in our versioning/release strategy is how to support major versions in parallel. I believe that when MiAIRR goes V2.0, we are still going to need to support the V1.0 toolset with bug fixes and maintenance.

**Comment by BusseChristian** [comment]
Sorry for being late to the party. I am pretty much fine with anything you guys decide. When I talked with Scott at AIRR2017, I had something related to *B* in mind. In such a model I still feel that PRs to `master` should be reviewed, as they will contain a larger set of changes and won't occur to often. In general I think that getting to a consensus on what the model is and what one should/should not do is the important step for now. Assuming that a "be considerate and know what you are doing" policy works, I would prefer it over a strict enforcement in the repo settings.
